### PR TITLE
CON-8179 : Onbaording AWS AWP

### PR DESCRIPTION
**Pull request by omartawfeeq**
This pull request introduces changes related to the AWP AWS onboarding process in the Dome9 Terraform provider.
Implementing AWP Onboarding API (https://github.com/dome9/dome9-sdk-go/releases/tag/v1.18.4)
* **resource** - dome9_awp_aws_onboarding
*  **data**  - dome9_awp_aws_onboarding
* **data**  -  dome9_awp_aws_onboarding_data

### CODE REVIEW

**Review by chkp-dolevh — COMMENTED** [review]
After the release of the SDK, the version that the provider uses needs to be updated

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: remove id
```suggestion
		arn    = "arn:aws:iam::<customer_account_id>:role/CloudGuard-Connect"
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: remove secret 
```suggestion
		secret = "<secret>"
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: add this:
```suggestion
// This is an example of how to use the Dome9 AWP AWS Onboarding Terraform provider.

## Poviders ##

# Required Providers Configuration Block for Dome9, AWS, HTTP, and Local
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf:41]: add
```suggestion
## CloudGuard Account Onboarding (Pre-requisite) ##
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: add
```suggestion
## CloudGuard AWP Configuration ##
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: chage:
```suggestion
	scan_mode = "inAccount|saas" # the valid values are "inAccount" and "saas" when onboarding the AWS account to Dome9 AWP.
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: add
```suggestion
## CloudGuard AWP Resources ##

// Note: count - Used as condition to create resources based on the scan mode.

```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: change:

```suggestion
	cloudguard_account_id = "dome9_cloudaccount_aws.aws_onboarding_account_test.id | <CLOUDGUARD_ACCOUNT_ID> | <EXTERNAL_AWS_ACCOUNT_NUMBER>"
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: this is  optional
```suggestion
# The dome9_awp_aws_onboarding data source allows you to get the onboarding data of an AWS account  (Optional).
```

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: the setting are optional, change that to this example:
```suggestion
	  # Optional Settings
  # e.g:
	#   agentless_account_settings { 					
	#     disabled_regions                 = []   // e.g: ["us-west-1", "us-west-2"]
	#     scan_machine_interval_in_hours   = 24
	#     max_concurrence_scans_per_region = 20
	#     skip_function_apps_scan          = false 
	#     custom_tags = {
	#       # example of custom tags
	#       tag1 = "value1"
	#       tag2 = "value2"
	#       tag3 = "value3"
	#     }
	#   }
```

**Review by batyakopelowitz — CHANGES_REQUESTED** [review]

File: examples/awp_aws_onboarding/main.tf
Comment (batyakopelowitz) [inline examples/awp_aws_onboarding/main.tf]: remove this env!!!
```suggestion
			
```

**Review by batyakopelowitz — CHANGES_REQUESTED** [review]

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf:41]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

File: examples/awp_aws_onboarding/main.tf
Comment (omartawfeeq) [reply examples/awp_aws_onboarding/main.tf]: done

**Review by batyakopelowitz — APPROVED** [review]

**Review by bzemms — CHANGES_REQUESTED** [review]
Wait for Alex approval

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (bzemms) [inline dome9/resource_dome9_awp_aws_onboarding.go]: this one is only for azure

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (bzemms) [inline dome9/resource_dome9_awp_aws_onboarding.go:56]: why its a list?

File: dome9/common/resourcetype/resource_type.go
Comment (chkp-alexgl) [inline dome9/common/resourcetype/resource_type.go]: remove "get". it's a data resource, not an operation.

File: dome9/common/testing/variable/variable.go
Comment (chkp-alexgl) [inline dome9/common/testing/variable/variable.go]: English: should be "concurrency" rather than "concurrence".

File: dome9/common/resourcetype/resource_type.go
Comment (omartawfeeq) [reply dome9/common/resourcetype/resource_type.go]: done

File: dome9/common/testing/variable/variable.go
Comment (omartawfeeq) [reply dome9/common/testing/variable/variable.go]: done

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go]: removing it

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go:56]: it is list with one item only, this is the only way that i can use list with mulktiple types, map with multiple types was not supported as i remember

File: dome9/common/resourcetype/resource_type.go
Comment (chkp-alexgl) [inline dome9/common/resourcetype/resource_type.go]: perhaps remove "Get" in variable name too, for consistency?

File: dome9/common/testing/environmentvariable/environment_variable.go
Comment (chkp-alexgl) [inline dome9/common/testing/environmentvariable/environment_variable.go]: we can't refer to internal SSMs in publicly available code

File: dome9/data_source_dome9_awp_aws_get_onboarding_data.go
Comment (chkp-alexgl) [inline dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: for consistency, I suggest removing "Get" from everywhere (.go file name in this case)

File: dome9/data_source_dome9_awp_aws_get_onboarding_data.go
Comment (chkp-alexgl) [inline dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: why do we need to retrieve all those details from backend - function name, runtime? Can't we just hard-code it in the module?

File: dome9/data_source_dome9_awp_aws_get_onboarding_data_test.go
Comment (chkp-alexgl) [inline dome9/data_source_dome9_awp_aws_get_onboarding_data_test.go]: remove "Get"

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (chkp-alexgl) [inline dome9/resource_dome9_awp_aws_onboarding.go:21]: do we support "update" of each of the individual parameters?

File: website/docs/d/awp_aws_onboarding_data.html.markdown
Comment (chkp-alexgl) [inline website/docs/d/awp_aws_onboarding_data.html.markdown]: remove "Get". Please, check all places.

File: dome9/common/resourcetype/resource_type.go
Comment (omartawfeeq) [reply dome9/common/resourcetype/resource_type.go]: Removing "Get" from the Variable name too

File: dome9/common/testing/environmentvariable/environment_variable.go
Comment (omartawfeeq) [reply dome9/common/testing/environmentvariable/environment_variable.go]: will removing it (@dolevh-d9 has not any objection for it (we need to refrence this ssm to be documented somewhere))

File: dome9/data_source_dome9_awp_aws_get_onboarding_data.go
Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: i am removing "Get" from everywhere, files and content

File: dome9/data_source_dome9_awp_aws_get_onboarding_data.go
Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: since when build the cft, we generate these variables, all of them need to be get from the api, since if these constants changed in awp agentless build process, we will also get the same change (so we decided to get all these values from the get onboarding data api)

File: dome9/data_source_dome9_awp_aws_get_onboarding_data_test.go
Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_get_onboarding_data_test.go]: done

File: website/docs/d/awp_aws_onboarding_data.html.markdown
Comment (omartawfeeq) [reply website/docs/d/awp_aws_onboarding_data.html.markdown]: no currently only for few including agentless_account_settings (since it has an update api)

File: dome9/data_source_dome9_awp_aws_get_onboarding_data.go
Comment (chkp-alexgl) [reply dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: @bzemms we need to keep in mind backwards compatibility -- customers might have existing deployments of these terraform resources, so we can't make breaking changes without careful consideration..

File: website/docs/d/awp_aws_onboarding_data.html.markdown
Comment (chkp-alexgl) [reply website/docs/d/awp_aws_onboarding_data.html.markdown]: so, what would happen if the user changes one of those other parameters (that we currently don't support changing) and runs 'terraform apply'?

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: i am writing here (the correct location) : 
no currently only for few including agentless_account_settings (since it has an update api)

File: website/docs/d/awp_aws_onboarding_data.html.markdown
Comment (omartawfeeq) [reply website/docs/d/awp_aws_onboarding_data.html.markdown]: nothing will be changed, once some other params changes that we don't support, terraform apply will not touch the resource or will not call any further apis (since we don't have any apis to change them), so it will not have any effect

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (chkp-alexgl) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: I think in some cases an "update" triggers removal and re-creation of the resource. Do we want to make it behave this way here too? Would it work? Alternatively, can we fail the update if the user attempts to change an unsupported field? If so, how the user would cause re-creation? Is there some flag in terraform for such cases?

File: dome9/data_source_dome9_awp_aws_onboarding.go
Comment (chkp-alexgl) [inline dome9/data_source_dome9_awp_aws_onboarding.go]: can we fix the typo and replace "max_concurre**nce**_scans" either with "max_concurre**ncy**_scans" or "max_concurre**nt**_scans" everywhere?

File: dome9/common/testing/environmentvariable/environment_variable.go
Comment (chkp-alexgl) [reply dome9/common/testing/environmentvariable/environment_variable.go]: why do we need (or want) the entire world to know about this SSM?

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (chkp-alexgl) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: Another option: reduce the parameters that the module will receive to only those that can be updated (not sure if it helps).

File: dome9/common/testing/environmentvariable/environment_variable.go
Comment (omartawfeeq) [reply dome9/common/testing/environmentvariable/environment_variable.go]: already removed the comment

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: we can't, since all of the needed parameters are required for the api call, without them we can't do proper api call

Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: I will attempt to implement when we try to update unsupported parameters, to fail the update flow
will check when the update triggers remocal and recreation and how if possible we can do recreation (to have the same behaviour)

File: dome9/data_source_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_onboarding.go]: changing to max_concurre**nt**_scans_per_region

Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_onboarding.go]: @bzemms we need also to change it in the module

Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_onboarding.go]: @bzemms will do it also on the module

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: @chkp-alexgl Regarding your questions :

1. in Terraform, an update can indeed trigger a removal and re-creation of a resource. This is typically controlled by the **ForceNew: true** property in the resource schema. **If this property is set to true for a field, then changing that field will cause the resource to be destroyed and recreated.**
2. in our resource schema, the fields cloudguard_account_id and scan_mode have the **ForceNew: true property**. This **means if these fields are changed in the Terraform configuration, the resource will be destroyed and recreated.**
3. i can failed the update and preserve the behaviour for other fields by : 
 using CustomizeDiff function in the resource schema. This function allows us to add custom logic to check for changes and return an error if necessary.
4. i will implement this, **since removal and recreation behavior for the fields with ForceNew: true will still be preserved.** This is because the ForceNew property is handled by Terraform itself and is not affected by the CustomizeDiff function.

Comment (chkp-alexgl) [reply dome9/resource_dome9_awp_aws_onboarding.go:21]: Great! @bzemms please, decide whether we prefer to fail the update, or to mark those fields as ForceNew which would trigger re-creation. I don't fully understand all the cases when the user would change these things to have an opinion.

File: dome9/data_source_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_onboarding.go]: @bzemms don't have permissions to update the module

File: dome9/data_source_dome9_awp_aws_onboarding.go
Comment (bzemms) [inline dome9/data_source_dome9_awp_aws_onboarding.go]: please remove

File: vendor/github.com/dome9/dome9-sdk-go/services/awp_aws_onboarding/awp_aws_onboarding.go
Comment (bzemms) [inline vendor/github.com/dome9/dome9-sdk-go/services/awp_aws_onboarding/awp_aws_onboarding.go:1]: ? this is SDK code or Dome9 Provider ?! I'm confused

File: vendor/github.com/dome9/dome9-sdk-go/services/awp_aws_onboarding/awp_aws_onboarding.go
Comment (bzemms) [inline vendor/github.com/dome9/dome9-sdk-go/services/awp_aws_onboarding/awp_aws_onboarding.go:81]: to remove

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (bzemms) [inline dome9/resource_dome9_awp_aws_onboarding.go]: SkipFunctionAppsScan - remove

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (bzemms) [inline dome9/resource_dome9_awp_aws_onboarding.go]: duplicated wrong condition

File: dome9/data_source_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/data_source_dome9_awp_aws_onboarding.go]: removing it, and remove any related logic for centralized (need to do the same validation when we want to support centralized onboarding)

File: vendor/github.com/dome9/dome9-sdk-go/services/awp_aws_onboarding/awp_aws_onboarding.go
Comment (omartawfeeq) [reply vendor/github.com/dome9/dome9-sdk-go/services/awp_aws_onboarding/awp_aws_onboarding.go:81]: currently this is an SDK effort, we can open another PR for any SDK change, otherwise, we need to update the sdk again and release it again

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go]: removing it

File: dome9/resource_dome9_awp_aws_onboarding.go
Comment (omartawfeeq) [reply dome9/resource_dome9_awp_aws_onboarding.go]: removing it

File: dome9/data_source_dome9_awp_aws_get_onboarding_data.go
Comment (bzemms) [reply dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: I don't get which resources. Do you mean if a customer already enabled AWP without TF ?

Comment (chkp-alexgl) [reply dome9/data_source_dome9_awp_aws_get_onboarding_data.go]: our lambda function. we can't really change its name or other parameters (in the future) - without worrying about backward compatibility with deployments before that future change. in fact, I see benefits in hard-coding it in the module rather than retrieving from backend. At least you can know which customer uses which version.

**Review by bzemms — APPROVED** [review]

**Review by chkp-dolevh — APPROVED** [review]

**Comment by chkp-alexgl** [comment]
there are still places that refer to "concurrence" and "concurrency". Do we want to change all to "concurrent" scans?

**Comment by omartawfeeq** [comment]
> there are still places that refer to "concurrence" and "concurrency". Do we want to change all to "concurrent" scans?

Let me check them now

**Comment by chkp-alexgl** [comment]
> there are still places that refer to "concurrence" and "concurrency". Do we want to change all to "concurrent" scans?

@bzemms I understand that there are some dependencies on SDK changes. Up to you whether to fix it now or in a separate PR. No other comments from me, thanks!

**Pull request merged**
